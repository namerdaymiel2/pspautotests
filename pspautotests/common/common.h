#undef main
#define main main2

extern int RUNNING_ON_EMULATOR;
