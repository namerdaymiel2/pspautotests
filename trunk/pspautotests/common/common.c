#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <pspdebug.h>
#include <pspthreadman.h>
#include <psploadexec.h>
#include <pspctrl.h>
#include <pspiofilemgr.h>
//#include <pspkdebug.h>

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <sys/lock.h>
#include <sys/fcntl.h>
//#include "local.h"

PSP_MODULE_INFO("TESTMODULE", PSP_MODULE_USER, 1, 0);
PSP_MAIN_THREAD_ATTR(THREAD_ATTR_USER | THREAD_ATTR_VFPU);

#define EMULATOR_DEVCTL__GET_HAS_DISPLAY 0x00000001

unsigned int RUNNING_ON_EMULATOR = 0;
unsigned int HAS_DISPLAY = 1;

extern int test_main(int argc, char *argv[]);

FILE stdout_back = {NULL};
SceUID KprintfFd = 0;

static int writeStdoutHook(struct _reent *ptr, void *cookie, const char *buf, int buf_len) {
	char temp[1024 + 1];

	if (KprintfFd != 0) sceIoWrite(KprintfFd, buf, buf_len);

	if (buf_len < sizeof(temp)) {
		if (HAS_DISPLAY) {
			memcpy(temp, buf, buf_len);
			temp[buf_len] = 0;

			//Kprintf("%s", temp);
			pspDebugScreenPrintf("%s", temp);
		}
	}
	
	if (stdout_back._write != NULL) {
		return stdout_back._write(ptr, cookie, buf, buf_len);
	} else {
		return buf_len;
	}
}

void test_begin() {
	if (HAS_DISPLAY) {
		pspDebugScreenInit();
	}

	if (RUNNING_ON_EMULATOR) {
		fclose(stdout);
		stdout = fmemopen(alloca(4), 4, "wb");
		stdout_back._write = NULL;
		
		//stderr = stdout;
		setbuf(stdout, NULL);
	} else {
		freopen("ms0:/__testoutput.txt", "wb", stdout);
		freopen("ms0:/__testerror.txt", "wb", stderr);
		stdout_back._write = stdout->_write;
	}
	stdout->_write = writeStdoutHook;
	
	setbuf(stderr, NULL);
}

void test_end() {
	fflush(stdout);
	fflush(stderr);

	if (!RUNNING_ON_EMULATOR) {
		SceCtrlData key;
		while (1) {
			sceCtrlReadBufferPositive(&key, 1);
			if (key.Buttons & PSP_CTRL_CROSS) break;
		}
	}
	
	//fclose(stdout);
	sceKernelExitGame();
	
	exit(0);
}

int test_psp_exit_callback(int arg1, int arg2, void *common) {
	exit(0);
	return 0;
}

int test_psp_callback_thread(SceSize args, void *argp) {
	int cbid;
	cbid = sceKernelCreateCallback("Exit Callback", test_psp_exit_callback, NULL);
	sceKernelRegisterExitCallback(cbid);
	sceKernelSleepThreadCB();
	return 0;
}

int test_psp_setup_callbacks(void) {
	int thid = 0;
	thid = sceKernelCreateThread("update_thread",  test_psp_callback_thread, 0x11, 0xFA0, 0, 0);
	if (thid >= 0) sceKernelStartThread(thid, 0, 0);
	return thid;
}

//#define START_WITH "ms0:/PSP/GAME/virtual"

/*
void emitInt(int v) {
	asm("syscall 0x1010");
}

void emitFloat(float v) {
	asm("syscall 0x1011");
}

void emitString(char *v) {
	asm("syscall 0x1012");
}

void emitComment(char *v) {
	asm("syscall 0x1012");
}

void emitMemoryBlock(void *address, unsigned int size) {
	asm("syscall 0x1013");
}

void emitHex(void *address, unsigned int size) {
	asm("syscall 0x1014");
}
*/

int main(int argc, char *argv[]) {
	int retval = 0;

	KprintfFd = sceIoOpen("emulator:/Kprintf", O_WRONLY, 0777);
	
	RUNNING_ON_EMULATOR = (KprintfFd > 0);

	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("emulator:", EMULATOR_DEVCTL__GET_HAS_DISPLAY, NULL, 0, &HAS_DISPLAY, sizeof(HAS_DISPLAY));
	}

	//if (strncmp(argv[0], START_WITH, strlen(START_WITH)) == 0) RUNNING_ON_EMULATOR = 1;

	if (!RUNNING_ON_EMULATOR) {
		test_psp_setup_callbacks();
	}
	atexit(sceKernelExitGame);

	test_begin();
	{
		pspDebugScreenPrintf("RUNNING_ON_EMULATOR: %s - %s\n", RUNNING_ON_EMULATOR ? "yes" : "no", argv[0]);
		retval = test_main(argc, argv);
	}
	test_end();
	
	return retval;
}
