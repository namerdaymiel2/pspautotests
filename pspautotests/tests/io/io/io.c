#include <common.h>

#include <pspkernel.h>
#include <pspctrl.h>
#include <pspdebug.h>
#include <pspdisplay.h>
#include <psptypes.h>
#include <pspiofilemgr.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/types.h>
#include <sys/unistd.h>

//#include "../common/emits.h"

char buf[MAXPATHLEN] = {0};
char startPath[MAXPATHLEN] = {0};

/**
 * Utility for .
 */
void checkChangePathsTry(const char *dest) {
	if (chdir(dest) < 0) {
		printf("(chdir error)\n");
	} else {
		printf("%s\n", getcwd(buf, MAXPATHLEN) ?: "(getcwd error)");
	}
}

/**
 * Check changing paths.
 */
void checkChangePaths() {
	printf("%s\n", getcwd(buf, MAXPATHLEN));
	checkChangePathsTry("");                 // empty string
	checkChangePathsTry("hello");            // nonexistent path
	checkChangePathsTry("..");               // parent dir
	checkChangePathsTry("../SAVEDATA");      // parent dir and subdir
	checkChangePathsTry("../..");            // multiple parents
	checkChangePathsTry(".");                // current dir
	checkChangePathsTry("./././//PSP");      // current dirs, extra slashes
	checkChangePathsTry("/PSP/./GAME");      // absolute with no drive
	checkChangePathsTry("/");                // root with no drive
	checkChangePathsTry("ms0:/PSP/GAME");    // absolute with drive
	checkChangePathsTry("flash0:/");         // different drive
	checkChangePathsTry("ms0:/PSP/../PSP/"); // mixed
	chdir(startPath);
}

/**
 * Check opening files.
 */
void checkFileOpen() {
	FILE *file = fopen("io.expected", "rb");
	printf("%d\n", file != NULL);
	if (file != NULL) {
		fseek(file, 0, SEEK_END);
		printf("%d\n", ftell(file) > 0);
		fclose(file);
	} else {
		printf("%d\n", 0);
	}
}

/**
 * Check listing directories.
 */
void checkDirectoryList() {
	DIR *dir;
	struct dirent *dp;
	dir = opendir(".");
	if (dir != NULL) {
		while ((dp = readdir(dir)) != NULL) {
			if (strncmp("io", dp->d_name, 2) == 0) {
				printf("%s\n", dp->d_name);
			}
		}
		closedir(dir);
	}
}

/**
 * Check listing directories extended.
 * Check that ms0:/PSP contains a directory called GAME with st_attr containing ONLY FIO_SO_IFDIR flag.
 */
void checkDirectoryListEx() {
	#define MAX_ENTRY 16
	SceIoDirent files[MAX_ENTRY];
	int fd, nfiles = 0;

	fd = sceIoDopen("ms0:/PSP");
	while (nfiles < MAX_ENTRY) {
		memset(&files[nfiles], 0x00, sizeof(SceIoDirent));
		if (sceIoDread(fd, &files[nfiles]) <= 0) break;
		if (files[nfiles].d_name[0] == '.') continue;

		if (strcmp(files[nfiles].d_name, "GAME") == 0) {
			// st_attr should only contain FIO_SO_IFDIR flag. (Required by NesterJ).
			if (files[nfiles].d_stat.st_attr == FIO_SO_IFDIR){
				printf("%s\n", files[nfiles].d_name);
				continue;
			}
		}
	}
	sceIoDclose(fd);
}

/**
 * Check that there are arguments passed to the app.
 * And the first argument should contain the path to the executable.
 * That path will be used to determine current working directory in libc/newlib.
 */
void checkMainArgs(int argc, char** argv) {
	int n;
	printf("%d\n", argc);
	for (n = 0; n < argc; n++) printf("%s\n", argv[n]);
}

int main(int argc, char** argv) {
	getcwd(startPath, MAXPATHLEN);
	checkMainArgs(argc, argv);
	checkChangePaths();
	checkFileOpen();
	checkDirectoryList();
	checkDirectoryListEx();
	return 0;
}