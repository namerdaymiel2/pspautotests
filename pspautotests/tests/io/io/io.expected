1
ms0:/PSP/GAME/TEST/EBOOT.PBP
ms0:/PSP/GAME/TEST
ms0:/PSP/GAME/TEST
(chdir error)
ms0:/PSP/GAME
ms0:/PSP/SAVEDATA
ms0:
ms0:
ms0:/PSP
ms0:/PSP/GAME
ms0:
ms0:/PSP/GAME
(chdir error)
ms0:/PSP
1
1
io.expected
GAME
